Stratis.Bitcoin/FullNode.cs

```diff
@@ -155,18 +157,125 @@ public void Start()
+	        using (var cts = new CancellationTokenSource())
+	        {
+	            Action shutdown = () =>
+	            {
+	                if (!cts.IsCancellationRequested)
+	                {
+	                    Console.WriteLine("Application is shutting down...");
+	                    try
+	                    {
+	                        cts.Cancel();
+	                    }
+	                    catch (ObjectDisposedException)
+	                    {
+	                    }
+	                }
+
+	                done.Wait();
+	            };
+
+                var assemblyLoadContext = AssemblyLoadContext.GetLoadContext(typeof(FullNode).GetTypeInfo().Assembly);
+                assemblyLoadContext.Unloading += context => shutdown();
+                Console.CancelKeyPress += (sender, eventArgs) =>
+	            {
+	                shutdown();
+	                // Don't terminate the process immediately, wait for the Main thread to exit gracefully.
+	                eventArgs.Cancel = true;
+	            };
+
+	            await this.RunAsync(cts.Token, "Application started. Press Ctrl+C to shut down.");
+	            done.Set();
+	        }
+        }
+
+	    public async Task RunAsync(CancellationToken cancellationToken, string shutdownMessage)
+	    {
+	        using (this)
+	        {
+	            await this.StartAsync(cancellationToken);
+
+	            if (!string.IsNullOrEmpty(shutdownMessage))
+	            {
+	                Console.WriteLine(shutdownMessage);
+	            }
+
+                cancellationToken.Register(state =>
+                    {
+                        ((IApplicationLifetime)state).StopApplication();
+                    },
+                    this._applicationLifetime);
+
+                var waitForStop = new TaskCompletionSource<object>(TaskCreationOptions.RunContinuationsAsynchronously);
+                this._applicationLifetime.ApplicationStopping.Register(obj =>
+                {
+                    var tcs = (TaskCompletionSource<object>)obj;
+                    tcs.TrySetResult(null);
+                }, waitForStop);
+
+                await waitForStop.Task;
+
+	            await StopAsync();
+	        }
+	    }
+
+	    public async Task StartAsync(CancellationToken cancellationToken)
+	    {
+	        await Task.Run(() => Start(), cancellationToken);
+	    }
+
+	    public async Task StopAsync()
+	    {
+	        await Task.Run(() => Stop(CancellationToken.None));
+	    }
+
+        public void Stop(CancellationToken cancellationToken)
+	    {
+	        if (this._stopped)
+	        {
+	            return;
+	        }
+            this._stopped = true;
+
+            // Fire IApplicationLifetime.Stopping
+            this._applicationLifetime?.StopApplication();
+
+	        if (this.GlobalCancellation != null)
+	        {
+                this.GlobalCancellation.Cancellation.Cancel();
+
+                this.ConnectionManager.Dispose();
```

Comment:
this.ConnectionManager.Dispose(); should probably go in the Base feature Stop() method